Allow different exception breakpoints from multiple debuggers to be shown at once

nisargjhaveri opened the pull request:
Fixes #126608.

Instead of replacing all exception breakpoints, add new exception breakpoints when a new debug sessions starts.

This way, the existing exception breakpoints stay as is. If no active debug session supports an existing exception breakpoint, it is shown as disabled and can be removed manually. A currently supported exception breakpoint cannot be removed manually as long as the session is active.

This gives more control to the user and saves the breakpoint preferences across different debug sessions in the workspace.

<!-- Thank you for submitting a Pull Request. Please:
* Read our Pull Request guidelines:
  https://github.com/microsoft/vscode/wiki/How-to-Contribute#pull-requests
* Associate an issue with the Pull Request.
* Ensure that the code is up-to-date with the `main` branch.
* Include a description of the proposed changes and how to test them.
-->

roblourens commented:
> Instead of replacing all exception breakpoints, add new exception breakpoints when a new debug sessions starts.

I don't think I like this approach. I would rather simply see the available exception breakpoint filters for the selected session. I think that mixing filters from different debuggers will be confusing. Also it seems that this PR allows for manually deleting exception breakpoint filters, and I don't understand that.

nisargjhaveri commented:
>Also it seems that this PR allows for manually deleting exception breakpoint filters, and I don't understand that.

With these changes, we're always adding new exception breakpoints, the old once (from other debuggers) are never automatically deleted. For that, I've added an option to delete exception breakpoints only if they're not from the current running debuggers.

>I don't think I like this approach. I would rather simply see the available exception breakpoint filters for the selected session. I think that mixing filters from different debuggers will be confusing.

Sure. Can you please suggest how should we approach this? I'm new and I could not figure out how we'll support storing preferences from previous runs, if we simply pull the available exception filters from the selected debugger. This seemed important as otherwise we'll always fallback to the defaults set by the adapter on starting the new session. Suggestions?

roblourens commented:
We should just show the exception filters for whichever debug session is active. It could be that a different model is loaded into the tree, or maybe a tree filter is helpful.

nisargjhaveri commented:
@roblourens I've updated the PR. Please have a look?

I've moved the exception breakpoints ownership to DebugSession. The exception breakpoints now can have per session enablement and other properties.

The model (and view) reflects exception breakpoints of the last focused session, it updates on "did focus" event. The enablement is restored/duplicated from model when a new debug session starts.

roblourens commented on src/vs/workbench/contrib/debug/browser/debugSession.ts:
```diff
@@ -302,7 +307,7 @@ export class DebugSession implements IDebugSession {
 			this.initialized = true;
 			this._onDidChangeState.fire();
-			this.debugService.setExceptionBreakpoints((this.raw && this.raw.capabilities.exceptionBreakpointFilters) || []);
+			this.exceptionBreakpoints = this.debugService.createExceptionBreakpoints((this.raw && this.raw.capabilities.exceptionBreakpointFilters) || []);
```
I don't think it's necessary to call all the way into the DebugModel here to create real ExceptionBreakpoint objects. I think it's ok for the DebugSession to expose raw exceptionBreakpointFilters and we can create real objects when that session is selected.

roblourens replied:
Seems that many things exposed by the DebugSession are raw DebugProtocol types

nisargjhaveri replied:
The alternative would be to keep a map from session to exception breakpoints in model or somewhere else, if we ultimately want different breakpoint configuration per session, no?

Theoretically, in this case, session owns the breakpoints, so made sense to keep it in session.

Model has the info required for creating the real ExceptionBreakpoint objects, i.e. values from last/current session to be used as default.

Agreed that the current flow is not intuitive. Suggestions?

roblourens replied:
I see that the enablement state of the exception breakpoint is on that object, so those need to live in some map somewhere. I think that putting them on the session is not right, because if I enable "caught exceptions" for one node debug session, that should be enabled for all node debug sessions. So you could have a map from debug adapter type to exception filters for that type.

But if I am debugging an electron app, you have node and chrome configs, those DAs contribute the same breakpoint filters, and if I enable "caught exceptions" for node, I want it to also be enabled for chrome. We could keep a set of all exception breakpoint filters each with the debug types that they are enabled for. I think we should actually be working towards something like that.

`ExceptionBreakpointsFilter` has an ID. If an unrelated DA contributes an exception filter with the same ID but a different label, should it be treated the same as the node/chrome filter? Or should we look at all fields of `ExceptionBreakpointsFilter` to decide whether it's a match?

nisargjhaveri replied:
> I see that the enablement state of the exception breakpoint is on that object, so those need to live in some map somewhere. I think that putting them on the session is not right, because if I enable "caught exceptions" for one node debug session, that should be enabled for all node debug sessions. So you could have a map from debug adapter type to exception filters for that type.

Having exception breakpoints in session, including enablement also fixes the scenario mentioned in #147096. But to be fair, if we want enablement different for every session, it could apply to other breakpoints as well, depending on the particular multi-session scenario. Though with this restoring earlier preferences doesn't work so well. Primarily because we'll anyway have to choose which session to store if multiple session from the same DA are active. To store different preferences for different DAs would also be challenging this case. 

> But if I am debugging an electron app, you have node and chrome configs, those DAs contribute the same breakpoint filters, and if I enable "caught exceptions" for node, I want it to also be enabled for chrome. We could keep a set of all exception breakpoint filters each with the debug types that they are enabled for. I think we should actually be working towards something like that.

> ExceptionBreakpointsFilter has an ID. If an unrelated DA contributes an exception filter with the same ID but a different label, should it be treated the same as the node/chrome filter? Or should we look at all fields of ExceptionBreakpointsFilter to decide whether it's a match?

What if we keep the exception breakpoints separate for each DA always, even if the id and name are same? We could show the breakpoints from all sessions, along with the DA name it is coming from to differentiate those. Which would also reduce confusion since the breakpoint pane won't change as per the active session. The user gets the choice to select exception breakpoints per DA. Since the two sessions are using different DAs, it is also likely that the significance and usage of the exception breakpoints can also be different, even if the id/name are same.

The original changes I made was somewhat similar to this, where the user sees all the exception breakpoints at once (though it did not differentiate per DA as such, it only looked at the id). The original changes also gave better experience when it comes to saving and restoring the breakpoints across multiple DAs, with a twist that if some EB from an inactive DA that are no longer needs to be saved, they needed to be manually deleted by the user. I feel something like that would give a much better experience when working with multiple debuggers at once. I can try this approach again if it makes sense.

roblourens replied:
> We could show the breakpoints from all sessions, along with the DA name it is coming from to differentiate those.

Strong disagree on that, that will be a lot of noise. It should act as a global view that updates for the active session, same as the debug toolbar.

> Since the two sessions are using different DAs, it is also likely that the significance and usage of the exception breakpoints can also be different, even if the id/name are same.

If this is about the node/chrome case, I disagree on that too. It's also sort of an issue that it's not necessarily obvious which sessions are "node" and which are "chrome" when debugging electron.

I think that the direction we are going in now is definitely the right one. We just need to decide whether exception breakpoint state is shared across DA types for the same ID (or ID + name, or something). I would like to say yes. It would be interesting to see whether any of the popular debug extensions have collisions in their exception filter IDs, I have no clue.

nisargjhaveri replied:
If we want to share EB state across sessions, session can no longer own it, which is the case with the current changes.

Here is what I'm thinking, though, cannot figure out a couple of things.

Model can keep a list of all exception breakpoints across currently active DAs. When a new session starts, we add exception breakpoints for it to the model if not already exists with same id/name. Model can then expose a method to get exception breakpoints for a particular session for the UI to display.

When do we remove the old exception breakpoints from the model in this case?

We could remove breakpoints from the model when the session ends, if no other active session supports it. If we do this, we regress the current behaviour of saving the enablement for exception breakpoints across restarts of a session.

We "could" never remove it. In that case the enablement is remembered even for multi-debugger cases as well, which is great. But the model would bloat over time in theory.

Btw, finding a design that can save enablement even for multi-debugger scenarios would help a lot. It is annoying that the exception breakpoints would fallback to default every time I start debugging with a multi-debugger setup.

Suggestions?

roblourens replied:
What you are describing is pretty much what I have in mind. I think that just looking at the ID is probably ok (worth taking a quick look at popular DA extensions) and saving the selection in workspace state forever is also fine.

nisargjhaveri replied:
Just looking at the id won't be sufficient since the breakpoints can have different support for conditions along with different names and descriptions.

For example, Java and Nodejs debugger has an overlapping id, but with different condition support and different descriptions.

I've updated the PR with changes, have a look?

roblourens replied:
> For example, Java and Nodejs debugger has an overlapping id, but with different condition support and different descriptions.

Interesting, what is that ID?

nisargjhaveri replied:
nodejs provides 'uncaught' and 'all', Java provides 'uncaught' and 'caught'. 'uncaught' is the conflicting one for these two.

Since the name are generic, I'd guess more debuggers would have similar conflicting names.

roblourens commented on src/vs/workbench/contrib/debug/browser/debugService.ts:
```diff
@@ -1026,9 +1027,16 @@ export class DebugService implements IDebugService {
-	setExceptionBreakpoints(data: DebugProtocol.ExceptionBreakpointsFilter[]): void {
-		this.model.setExceptionBreakpoints(data);
-		this.debugStorage.storeBreakpoints(this.model);
+	updateExceptionBreakpoints(): void {
```
Is this supposed to be public?

nisargjhaveri replied:
Nope. Will update.

roblourens commented on src/vs/workbench/contrib/debug/common/debugModel.ts:
```diff
@@ -1436,7 +1436,8 @@ export class DebugModel implements IDebugModel {
 		});
-		this.exceptionBreakpoints.forEach(ebp => {
+		this.exceptionBreakpoints.forEach(iebp => {
+			const ebp = iebp as ExceptionBreakpoint;
```
It's ok for `this.exceptionBreakpoints` to be the real `ExceptionBreakpoint` class, that seems to be the case for other properties

nisargjhaveri replied:
The real ExceptionBreakpoint is defined in model and though exported, is generally not used in debug service or session. Since currently model creates ExceptionBreakpoint objects, and only model or view needs to use these objects. Session is only holding IExceptionBreakpoint.

With the current design, it is either a "type cast" here or in setExceptionBreakpoint method. Open to suggestions if we should be doing something else.

roblourens commented:
A couple comments but this is basically what I had in mind, thank you for the PR!

nisargjhaveri commented:
Updated the PR.

Some behaviour differences with this:
- The exception breakpoints shows up when the session is done attaching, compared to when the session was started earlier.
- When no session is active, no exception breakpoints are visible. Earlier, we kept showing the once from the last session.

nisargjhaveri commented on src/vs/workbench/contrib/debug/common/debugModel.ts:
```diff
@@ -1340,23 +1359,29 @@ export class DebugModel implements IDebugModel {
 		return this.exceptionBreakpoints;
 	}
 
+	getExceptionBreakpointsForSession(sessionId?: string): IExceptionBreakpoint[] {
```
I've changes some usages of `getExceptionBreakpoints` with the new `getExceptionBreakpointsForSession` when appropriate.

The usage in `logDebugSessionStart` in `debugTelemetry.ts`, I'm not sure about what is the preferred one to use.

roblourens commented:
> When no session is active, no exception breakpoints are visible. Earlier, we kept showing the once from the last session.

I actually think it should continue to show the breakpoints from the last session. I think that the most common usecase is rerunning the same debug type, and changing these exception breakpoints before you start debugging is a common task.

nisargjhaveri commented:
> I actually think it should continue to show the breakpoints from the last session. I think that the most common usecase is rerunning the same debug type, and changing these exception breakpoints before you start debugging is a common task.

Agreed. I've pushed one more change to fix this.

nisargjhaveri commented:
@roblourens is anything pending for this one? It'd be great if you could have a look once at the latest changes.

roblourens commented:
Sorry, I didn't have time to work on it this month but I'll try to take a look again after our release next week.

roblourens commented on src/vs/workbench/contrib/debug/common/debugModel.ts:
```diff
@@ -1075,14 +1078,36 @@ export class ExceptionBreakpoint extends BaseBreakpoint implements IExceptionBre
+	}
+
+	setFallback(isFallback: boolean) {
+		this.fallback = isFallback;
```
Seems like there should be a nicer way to accomplish this. Seems like `fallback` is not a concept that an individual exception breakpoint should be concerned with. But I'm not sure what the solution is, and I am ok with keeping it as-is. If this is kept, please add a comment describing what `fallback` means, it wasn't immediately clear.

nisargjhaveri replied:
I could not find a better way, especially which allows for easily persisting this as well. I've added some comments, which should help understanding the need for this.

roblourens reviewed (commented):
This works great, thanks for tackling it. Last request- could you think about testing for the new behavior? A couple tests in breakpoints.test.ts would probably be sufficient.

nisargjhaveri commented:
@roblourens I've updated the review and added more tests for this as well, which should cover most scenarios.
